application/cms/views.py

```diff
@@ -1109,3 +1112,167 @@ def view_measure_version_by_measure_version_id(measure_version_id):
+        db.session.commit()
+
+        return redirect(
+            url_for(
+                "cms.edit_measure_version",
+                topic_slug=topic.slug,
+                subtopic_slug=subtopic.slug,
+                measure_slug=measure.slug,
+                version=measure_version.version,
+            )
+        )
+
+    else:
+
+        errors = get_form_errors(forms=[data_source_form])
+
+        # TODO: delete this once the validation has been added to the form.
+        errors.append(ErrorSummaryMessage(href="#title-label", text="Enter a title"))
+
+        return render_template(
+            "cms/new_data_source.html",
+            data_source_form=data_source_form,
+            organisations_by_type=Organisation.select_options_by_type(),
+            errors=errors,
+            topic=topic,
+            subtopic=subtopic,
+            measure=measure,
+            measure_version=measure_version,
+        )
+
+
+@cms_blueprint.route(  # noqa: C901 (complexity)
+    "/<topic_slug>/<subtopic_slug>/<measure_slug>/<version>/edit/data-sources/<data_source_id>", methods=["GET"]
+)
+@login_required
+@user_has_access
+def edit_data_source(topic_slug, subtopic_slug, measure_slug, version, data_source_id):
+
+    topic, subtopic, measure, measure_version = page_service.get_measure_version_hierarchy(
+        topic_slug, subtopic_slug, measure_slug, version
+    )
+
+    data_source = DataSource.query.get(data_source_id)
+
+    if data_source is None:
+        raise PageNotFoundException()
+
+    if data_source not in measure_version.data_sources:
+        raise PageNotFoundException()
+
+    data_source_form = DataSourceForm(obj=data_source)
+    organisations_by_type = Organisation.select_options_by_type()
+
+    return render_template(
+        "cms/edit_data_source.html",
+        data_source=data_source,
+        data_source_form=data_source_form,
+        organisations_by_type=organisations_by_type,
+        topic=topic,
+        subtopic=subtopic,
+        measure=measure,
+        measure_version=measure_version,
+    )
+
+
+@cms_blueprint.route(  # noqa: C901 (complexity)
+    "/<topic_slug>/<subtopic_slug>/<measure_slug>/<version>/edit/data-sources/<data_source_id>", methods=["POST"]
+)
+@login_required
+@user_has_access
+def update_data_source(topic_slug, subtopic_slug, measure_slug, version, data_source_id):
+
+    topic, subtopic, measure, measure_version = page_service.get_measure_version_hierarchy(
+        topic_slug, subtopic_slug, measure_slug, version
+    )
+
+    data_source = DataSource.query.get(data_source_id)
+
+    if data_source is None:
+        raise PageNotFoundException()
+
+    data_source_form = DataSourceForm(sending_to_review=False)
+    data_source_form.populate_obj(data_source)
+
+    # TODO: remove check on data_source.title and validate this within form.
+    if data_source_form.validate_on_submit() and data_source.title:
+
+        db.session.commit()
+
+        return redirect(
```

Comment:
It would be good to flash a success message for users to confirm that their changes have been saved here.

Reply:
I tend to think that redirecting to a different page is enough of a confirmation – as if it didn't save, you'd see an error?

Reply:
It definitely felt confusing to me. It isn't consistent with existing behaviour, which keeps you on the same page (saving a measure, saving a dimension, saving a chart/table). Maybe it should just keep you on the same page.

But if we keep it as a redirect back to the measure, I feel like we definitely need to acknowledge that's what has happened.

Reply:
Hmm, well originally it kept your on the page, but we decided it should take you to the measure page...

Can add a success message for consistency, although personally I'm not a fan of them.